pkg/pbq/store/wal/bootstrap.go

```diff
@@ -0,0 +1,198 @@
+	"encoding/binary"
+	"encoding/gob"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/numaproj/numaflow/pkg/isb"
+	"github.com/numaproj/numaflow/pkg/pbq/partition"
+)
+
+var location *time.Location
+
+func init() {
+	var err error
+	location, err = time.LoadLocation("UTC")
+	if err != nil {
+		panic(fmt.Sprint("cannot load UTC", err))
+	}
+}
+
+// IsCorrupted checks whether the file is corrupt
+func (w *WAL) IsCorrupted() bool {
+	return w.corrupted
+}
+
+// OpenWAL returns a WAL if present
+func OpenWAL(filePath string, o *options) (*WAL, error) {
+	stat, err := os.Stat(filePath)
+	// there could be other errors
+	if os.IsNotExist(err) {
+		return nil, nil
+	} else if err != nil {
+		return nil, err
+	}
+
+	// here we are explicitly giving O_RDWR because we will be using this to read too. Our read is only during
+	// boot up.
+	fp, err := os.OpenFile(filePath, os.O_RDWR|os.O_APPEND, stat.Mode())
+	if err != nil {
+		return nil, err
+	}
+
+	w := &WAL{
+		fp:       fp,
+		openMode: os.O_RDWR,
+	}
+
+	w.partitionID, err = w.readHeader()
+
+	return w, err
+}
+
+func (w *WAL) readHeader() (*partition.ID, error) {
+	if w.openMode == os.O_WRONLY {
+		return nil, fmt.Errorf("opened using O_WRONLY")
+	}
+
+	if w.rOffset > 0 {
+		return nil, fmt.Errorf("header has already been read, current readoffset is at %d", w.rOffset)
+	}
+
+	id, err := decodeHeader(w.fp)
+	if err != nil {
+		return nil, err
+	}
+
+	seek, err := w.fp.Seek(0, io.SeekCurrent)
+	if err != nil {
+		return nil, err
+	}
+
+	w.rOffset += seek
+
+	return id, err
+}
+
+func (w *WAL) isEnd() bool {
+	// TODO: If we are done reading, check that we got the expected
+	// number of entries and return EOF.
+	return w.rOffset >= w.readUpTo
+}
+
+// decodeHeader decodes the header which is encoded by encodeHeader.
+func decodeHeader(buf io.Reader) (*partition.ID, error) {
+	var err error
+	// read the fixed vals
+	var hp = new(headerPreamble)
+	err = binary.Read(buf, binary.LittleEndian, hp)
+	if err != nil {
+		return nil, err
+	}
+	// read the variadic key
+	var key = make([]rune, hp.KLen)
+	err = binary.Read(buf, binary.LittleEndian, key)
+	if err != nil {
+		return nil, err
+	}
+
+	return &partition.ID{
+		Start: time.UnixMilli(hp.S).In(location),
+		End:   time.UnixMilli(hp.E).In(location),
+		Key:   string(key),
+	}, nil
+}
+
+func (w *WAL) Read(size int64) ([]*isb.ReadMessage, bool, error) {
+	if w.openMode == os.O_WRONLY {
+		return nil, false, fmt.Errorf("opened using O_WRONLY")
+	}
+
+	messages := make([]*isb.ReadMessage, 0)
+	// if size is greater than the number of messages in the store
+	// we will assign size with the number of messages in the store
+	start := w.rOffset
+	for size > w.rOffset-start && !w.isEnd() {
+		message, sizeRead, err := decodeEntry(w.fp)
+		if err != nil {
+			if err == errChecksumMismatch {
+				w.corrupted = true
+			}
+			return nil, false, err
+		}
+
+		w.rOffset += sizeRead
+		messages = append(messages, message)
+	}
+
+	if w.isEnd() {
+		w.wOffset = w.rOffset
```

Comment:
this could potentially become a bug. If we start writing after the read during bootup has finished, we will wrongly assume that unless the write and read offsets match we have not finished reading.

Reply:
yes, but I am suspecting there could be some corner case unless we might reissue a `Read` even if the read has been completed. Perhaps use a flag to return an error if such a read is attempted?

Reply:
yeah I can have a check that when a Read start, it always start from bootstrap not any other cases.